demos/cancel/README.md

```diff
@@ -0,0 +1,36 @@
+# Cancel
+
+In this demo we intend to show that we can use a syscall to terminate a thread and release the resources that used to own by it.
+
+To illustrate the mechanism, we must introduce the following:
+
+* `thread_wait` (in `src/user/thread.c`)
+* `thread_cancel` (in `src/user/thread.c`)
+
+## `thread_wait`
+
+This function sets the state of the current thread to `wait`.
+
+Since the scheduler will not pick a thread in `wait` state, the thread will keep blocking until we change its state.
+
+## `thread_cancel`
+
+When we call this function, it will cancel the thread specified by the `tid`.
+
+If `tid` is given as `CURRENT_THREAD`, the current thread will be terminated. The scheduler will choose another thread to run.
+
+If we apply `cancel()` to a valid thread, the memory space used it used will be cleaned up. Then we can create another thread on the released memory space.
+
+When we call the function to cancel an invalid thread, it will return `false`. No thread will be terminated.
+
+## Walkthrough
+
+In the beginning, the `setup()` function creates a thread called `cancel_self` which cancels itself after it starts to run.
+
+When the scheduler chooses `setup` thread to run again, it creates another 3 threads and set itself to `wait` state. This is to occupy the memory space for the thread structure, which can ensure the operation later.
+
+Then `canceller` thread cancelled thread 3 and adds a new thread. Therefore we will place the new thread into space that used to store the cancelled thread before. As a result, the `tid` of the new thread will be `3`.
+
+After that, the `canceller` thread terminated both thread 3 and thread 1. Since thread 1 has been cancelled, when we use `thread_join()` to join thread 1, we will see that its state is `cancelled`.
+
+Finally, the `canceller` cancels thread 0 and ends the demo.
```

Comment:
Here you can reference that 0 is waiting.

"Finally, the `canceller` cancels thread 0 (which is still in the `waiting` state) and returns. The demo now ends because there are no valid threads left."